### PR TITLE
Feasture/graph builder

### PR DESCRIPTION
**Pull request by volkale**

### CODE REVIEW

File: bonspy/graph_builder.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    def _get_lazy_formatter(formatters):
+        if not formatters:
+            return defaultdict(lambda: (lambda x: x))
+        else:
+            lazy_formatters = defaultdict(lambda: str)
+            lazy_formatters.update(formatters)
+            return lazy_formatters
+
+    def get_data(self):
+        for file in self.input_:
```
Comment (waltherg) [inline bonspy/graph_builder.py:39]: reserved keyword

File: bonspy/graph_builder.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        self.features = features
+        self.types_iterable = self._get_types_iterable(types_dict)
+        self.lazy_formatters = self._get_lazy_formatter(lazy_formatters)
+        self.functions = functions
+
+    def _get_types_iterable(self, types_dict):
+        return tuple(types_dict.get(f, 'assignment') for f in self.features)
+
+    @staticmethod
+    def _get_lazy_formatter(formatters):
```
Comment (waltherg) [inline bonspy/graph_builder.py:30]: accept formatters as dict to begin with?

File: bonspy/graph_builder.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        except StopIteration:
+            return True
+
+    def _get_state(self, graph, parent, new_feature=None):
+        state = graph.node[parent]['state'].copy()
+        new_state = self._add_new_feature(state, new_feature) if new_feature else state
+        return new_state
+
+    @staticmethod
+    def _add_node(graph, new_node, state, **kwargs):
```
Comment (waltherg) [inline bonspy/graph_builder.py]: method not necessary?

File: bonspy/graph_builder.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        return graph, node_index
+
+    def _add_branch(self, graph, row, node_index):
+        parent = 0
+        graph.node[parent] = self._apply_functions(graph.node[parent], row)
+
+        for feature in self.features:
+            feature_value = row[feature]
+            child = self._get_child(graph, parent, feature, feature_value)
+            if not child:
```
Comment (waltherg) [inline bonspy/graph_builder.py]: is  not None

File: bonspy/graph_builder.py (hunk 2 above)
Comment (volkale) [reply bonspy/graph_builder.py:30]: code works for both input types, but will update docstrings

**Review by waltherg — APPROVED** [review]

**Pull request merged**